### PR TITLE
Improvements in lost replica recovery

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed too low default value of `max_replicated_logs_to_keep` setting, which might cause replicas to become lost too often. Improve lost replica recovery process by choosing the most up-to-date replica to clone. Also do not remove old parts from lost replica, detach them instead.

### CODE REVIEW

**Comment by tavplubix** [comment]
Integration tests (release) - test_always_fetch_merged/test.py::test_replica_always_download - flacky
Performance - unrelated
Stress test (undefined) — Server failed to start - not a bug, seems like we need more retries in alive check

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -2336,47 +2345,107 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
...
-            if (!zookeeper->tryGet(source_replica_path + "/is_lost", source_replica_is_lost_value, &source_is_lost_stat)
-                || source_replica_is_lost_value == "0")
-            {
-                source_replica = source_replica_name;
-                break;
-            }
+            /// For compatibility with older ClickHouse versions
+            get_is_lost.stat.version = -1;
+        }
+        else if (get_is_lost.data != "0")
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:2419]: Write a log message about it?

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -2336,47 +2345,107 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
...
+    /// Wait for results before getting log entries
+    for (auto & future : futures)
+        future.wait();
+
+    Strings log_entries = zookeeper->getChildren(zookeeper_path + "/log");
+    size_t max_log_entry = 0;
+    if (!log_entries.empty())
+    {
+        String last_entry_num = std::max_element(log_entries.begin(), log_entries.end())->substr(4);
+        max_log_entry = std::stol(last_entry_num);
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:2396]: ` parse<UInt64>(last_entry_num.substr(strlen("log-")))` looks more clear

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -2336,47 +2345,107 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
...
+            source_replica = source_replica_name;
+            source_is_lost_stat = get_is_lost.stat;
+            min_replication_lag = replication_lag;
         }
     }
 
     if (source_replica.empty())
         throw Exception("All replicas are lost", ErrorCodes::ALL_REPLICAS_LOST);
 
+    if (is_new_replica)
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:2444]: let's also add info about source queue size and log pointer.

**Review by alesapin — APPROVED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 3 above)
Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:2444]: Maybe add log pointer here?
https://github.com/ClickHouse/ClickHouse/blob/47e9fe9ff16aa48c3f9b77ea1e9764b9cc17af20/src/Storages/StorageReplicatedMergeTree.cpp#L2432

**Comment by tavplubix** [comment]
AST fuzzer - known issue
Docs check - unrelated
Performance - unrelated

**Pull request merged**